component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/voiceotp/VoiceOTPAuthenticator.java

```diff
@@ -0,0 +1,2034 @@
+                        throw new AuthenticationFailedException("Mobile claim update failed for user :" + username, e);
+                    } catch (UserStoreException e) {
+                        Throwable ex = e.getCause();
+                        if (ex instanceof UserStoreClientException) {
+                            context.setProperty(VoiceOTPConstants.MOBILE_NUMBER_UPDATE_FAILURE, "true");
+                            context.setProperty(VoiceOTPConstants.PROFILE_UPDATE_FAILURE_REASON, ex.getMessage());
+                        }
+                        throw new AuthenticationFailedException("Mobile claim update failed for user " + username, e);
+                    }
+                }
+            }
+        }
+
+        if (!succeededAttempt) {
+            handleVoiceOtpVerificationFail(context);
+            context.setProperty(VoiceOTPConstants.CODE_MISMATCH, true);
+            throw new AuthenticationFailedException("Invalid code. Verification failed.");
+        }
+        // It reached here means the authentication was successful.
+        resetVoiceOtpFailedAttempts(context);
+    }
+
+    /**
+     * This method will process the otp entered is valid.
+     *
+     * @param context
+     * @param authenticatedUser
+     * @throws AuthenticationFailedException
+     */
+    private void processValidUserToken(AuthenticationContext context, AuthenticatedUser authenticatedUser) throws
+            AuthenticationFailedException {
+
+        Optional<Object> tokenValidityTime = Optional.ofNullable(context.getProperty(VoiceOTPConstants.
+                TOKEN_VALIDITY_TIME));
+        if (!tokenValidityTime.isPresent() || !NumberUtils.isNumber(tokenValidityTime.get().toString())) {
+            log.error("TokenExpiryTime property is not configured in application-authentication.xml or Voice OTP " +
+                    "Authenticator UI");
+            context.setSubject(authenticatedUser);
+            return;
+        }
+
+        Optional<Object> otpTokenSentTime = Optional.ofNullable(context.getProperty(VoiceOTPConstants.
+                SENT_OTP_TOKEN_TIME));
+        if (!otpTokenSentTime.isPresent() || !NumberUtils.isNumber(otpTokenSentTime.get().toString())) {
+
+            log.debug("Could not find OTP sent time");
+
+            throw new AuthenticationFailedException("Internal Error Occurred");
+        }
+
+        long elapsedTokenTime = System.currentTimeMillis() - Long.parseLong(otpTokenSentTime.get().toString());
+
+        if (elapsedTokenTime <= (Long.parseLong(tokenValidityTime.get().toString()) * 1000)) {
+            context.removeProperty(VoiceOTPConstants.TOKEN_EXPIRED);
+            context.setSubject(authenticatedUser);
+        } else {
+            context.setProperty(VoiceOTPConstants.TOKEN_EXPIRED, VoiceOTPConstants.TOKEN_EXPIRED_VALUE);
+            handleVoiceOtpVerificationFail(context);
+            throw new AuthenticationFailedException("OTP code has expired");
```

Comment:
Ideally these errors should be defined as a constant with an error code and that predefined error should be referred from the places where it is thrown. That will make error management easy and consistent throughout the system.

Reply:
Had a discussion about this and as the earlier discussed implementation is already done,we will create a git issue for this to do task on this repo to track this in future.

Reply:
https://github.com/wso2-extensions/identity-outbound-auth-voice-otp/issues/2